### DISCUSSION TITLE
module has no exported member

### DISCUSSION THREAD

**Discussion by dagda1** [pnpm/pnpm#4331] [Q&A] [answered]
I would really appreciate if somebody could explain why pnpm is not importing the dependencies of a module that I am importing.

I am importing [effection@2.0.4](https://github.com/thefrontside/effection) by having the following in my package.json file:

```json
"dependencies": {
  "effection": "^2.0.4",
```

Effection's package.json file looks like this:

```json
{
  "name": "effection",
  "version": "2.0.4",
  "sideEffects": false,
  "main": "dist-cjs/index.js",
  "module": "dist-esm/index.js",
  "types": "dist-esm/index.d.ts",
  "dependencies": {
    "@effection/channel": "2.0.3",
    "@effection/core": "2.2.0",
    "@effection/events": "2.0.3",
    "@effection/fetch": "2.0.4",
    "@effection/main": "2.0.3",
    "@effection/subscription": "2.0.3",
    "@effection/stream": "2.0.3"
  },
  ```
The module file `dist-esm/index.js` file of effection does an `import * ` from the modules listed in the `dependencies` above
  
```ts
export * from '@effection/core';
export * from '@effection/main';
export * from '@effection/channel';
export * from '@effection/events';
export * from '@effection/fetch';
export * from '@effection/subscription';
export * from '@effection/stream';
```

But pnpm does not have any symlinks in my `node_modules` for `@effection`,  nor is there an `@effection` folder in `node_modules/effection/node_modules`

Can anyone please suggest why this would be?

This previously worked with yarn.

**Comment by zkochan** [comment]
`node_modules/effection` is a symlink to `node_modules/.pnpm/effection@2.0.4/node_modules/effection`

The dependencies of effection are at `node_modules/.pnpm/effection@2.0.4/node_modules`.

This works because node.js looks for dependencies at the real location of the file.

See https://pnpm.io/symlinked-node-modules-structure

**Reply by dagda1** [reply]
i think the problem is that the dependencies of `@effection/*` were not downloaded

**Comment by dagda1** [comment, marked as answer]
@zkochan I found out what the problem was but I'm not entirely sure why.

It was the typescript types  or declaration files that were not resolving and it was because `preserveSymlinks` was in the `tsconfig.json` which was set to `true`.

removing the option makes everything work.

Do you know why this would cause it to fail?  I thought (wrongly) `preserveSymlinks` would be essential for making pnpm work in a monorepo.

**Reply by Pagebakers** [reply]
Only setting shamefully-hoist to true resolves it properly for me.

https://pnpm.io/npmrc#shamefully-hoist

**Reply by zkochan** [reply]
no, shamefully-hoist is not a solution for this problem. `preserveSymlinks` should never be `true` with pnpm. But with other package managers it is also a bad idea to set it to true. 

Alternatively, you can use `node-linker=hoisted`. But that is also just a workaround. The right solution is preserveSymlinks=false

**Reply by Mathdrquinn** [reply]
Thank you all for this helpful thread.  I'm trying the above proposed solutions and they are not working with my PNPM workspace.  I submitted an [issue to Vite](https://github.com/vitejs/vite/issues/11657), but maybe this is a good place to raise it up.

Here's my reproduction.  Running `pnpm --filter app1 build` fails, sadly: https://stackblitz.com/edit/vitejs-vite-jgvxwy?file=.npmrc

Any insights you have would be greatly appreciated!

**Reply by aestheticsdata** [reply]
> no, shamefully-hoist is not a solution for this problem. `preserveSymlinks` should never be `true` with pnpm. But with other package managers it is also a bad idea to set it to true.
> 
> Alternatively, you can use `node-linker=hoisted`. But that is also just a workaround. The right solution is preserveSymlinks=false

ok but why `preserveSymlinks` should be set to `false` with pnpm ? I have installed a lot npm modules for one year without any issue with pnpm in a monorepo with `preserveSymLinks` set to `true`.
I have this issue with tanstack-query and pnpm in a monorepo. It works if I set `preserveSymLinks` to `false`, but then I have multiple typescript errors with others libs (magic-link, html-react-parser).
I opened an issue on the github of tanstack : https://github.com/orgs/pnpm/discussions/4331#discussioncomment-3379386
